column_test.go

```diff
@@ -397,3 +410,110 @@ func (i *fileOffsetIndex) CompressedPageSize(j int) int64 {
 	return int64(i.PageLocations[j].CompressedPageSize)
 }
 func (i *fileOffsetIndex) FirstRowIndex(j int) int64 { return i.PageLocations[j].FirstRowIndex }
+
+type columnStats struct {
+	page       int
+	columnType parquet.Type
+	minValues  []parquet.Value
+	maxValues  []parquet.Value
+}
+
+func newColumnStats(columnType parquet.Type) *columnStats {
+	return &columnStats{columnType: columnType}
+}
+
+func (c *columnStats) observe(value parquet.Value) {
+	if c.page >= len(c.minValues) {
+		c.minValues = append(c.minValues, value)
+	} else if c.columnType.Compare(c.minValues[c.page], value) > 0 {
+		c.minValues[c.page] = value
+	}
+
+	if c.page >= len(c.maxValues) {
+		c.maxValues = append(c.maxValues, value)
+	} else if c.columnType.Compare(c.maxValues[c.page], value) < 0 {
+		c.maxValues[c.page] = value
+	}
+}
+
+func (c *columnStats) pageRead() {
+	c.page++
+}
+
+type indexOrder int
+
+const (
+	invalidIndexOrder indexOrder = iota
+	unorderedIndexOrder
+	ascendingIndexOrder
+	descendingIndexOrder
+)
+
+func (o indexOrder) String() string {
+	switch o {
+	case unorderedIndexOrder:
+		return "unordered"
+	case ascendingIndexOrder:
+		return "ascending"
+	case descendingIndexOrder:
+		return "descending"
+	default:
+		return "invalid"
+	}
+}
+
+func columnIndexOrder(index parquet.ColumnIndex) indexOrder {
+	switch {
+	case index.IsAscending() && index.IsDescending():
+		return invalidIndexOrder
+	case index.IsAscending():
+		return ascendingIndexOrder
+	case index.IsDescending():
+		return descendingIndexOrder
+	default:
+		return unorderedIndexOrder
+	}
+}
+
+func observedIndexOrder(columnType parquet.Type, minValues []parquet.Value, maxValues []parquet.Value) indexOrder {
+	a := valueOrder(columnType, minValues)
+	b := valueOrder(columnType, maxValues)
+
+	switch {
+	case a == ascendingIndexOrder && b == ascendingIndexOrder:
+		return ascendingIndexOrder
+	case a == descendingIndexOrder && b == descendingIndexOrder:
+		return descendingIndexOrder
+	default:
+		return unorderedIndexOrder
+	}
+}
+
+func valueOrder(columnType parquet.Type, values []parquet.Value) indexOrder {
+	switch len(values) {
+	case 0, 1:
+		return unorderedIndexOrder
+	}
+
+	var order int
+	for i := 1; i < len(values); i++ {
+		next := columnType.Compare(values[i], values[i-1])
+		if next == 0 {
+			continue
+		}
+		if order == 0 {
+			order = next
+			continue
+		}
+		if order != next {
+			return unorderedIndexOrder
+		}
+	}
+
+	switch order {
+	case -1:
```

Comment:
The `parquet.Type.Compare` method promises that the values will be negative, zero, or positive, but not equal to -1, 0, or +1 https://pkg.go.dev/github.com/segmentio/parquet-go#Type

How about using `< 0` instead here?

`parquet.Type.Compare` will return a negative value if the first value is less than the second, in which case the order is ascending as well isn't it? It seems this logic is inverted then?

Reply:
> `parquet.Type.Compare` will return a negative value if the first value is less than the second, in which case the order is ascending as well isn't it? It seems this logic is inverted then?

That's correct. This code is calling [`Compare(second, first)`](https://github.com/segmentio/parquet-go/blob/1b616d0ef40d6ac4d55c1448c582dd4edd6980d7/column_test.go#L500) because I think in a previous version I was comparing it to `OrderBy*()`. That's confusing so I updated the code to do what's expected.

Fixed that and better adherence to the `parquet.Type.Compare` API in 1aa9ce4

Reply:
Thanks for the clarification, it was confusing indeed to have `Compare(second, first)` 👍